backends/tc/backend.cpp

```diff
@@ -0,0 +1,623 @@
+                for (auto anno : annoList) {
+                    if (anno->name != "tc_type") continue;
+                    auto expressionList = anno->expr;
+                    for (auto expr : expressionList) {
+                        if (auto typeLiteral = expr->to<IR::StringLiteral>()) {
+                            auto val = checkTcType(typeLiteral);
+                            if (val != BIT_TYPE) {
+                                tcActionParam->setDataType(val);
+                            } else {
+                                ::error(ErrorType::ERR_INVALID,
+                                        "tc_type annotation cannot have '%1%' as value", expr);
+                            }
+                        } else {
+                            ::error(ErrorType::ERR_INVALID,
+                                    "tc_type annotation cannot have '%1%' as value", expr);
+                        }
+                    }
+                }
+                tcAction->addActionParams(tcActionParam);
+            }
+        }
+        tcPipeline->addActionDefinition(tcAction);
+    }
+}
+
+void ConvertToBackendIR::updateDefaultMissAction(const IR::P4Table *t, IR::TCTable *tabledef) {
+    auto defaultAction = t->getDefaultAction();
+    if (defaultAction == nullptr || !defaultAction->is<IR::MethodCallExpression>()) return;
+    auto methodexp = defaultAction->to<IR::MethodCallExpression>();
+    if ((methodexp == nullptr) || (methodexp->method == nullptr) ||
+        (!methodexp->method->is<IR::PathExpression>()))
+        return;
+    auto pathexp = methodexp->method->to<IR::PathExpression>();
+    if (pathexp == nullptr || pathexp->path == nullptr) return;
+    auto actionDecl = refMap->getDeclaration(pathexp->path);
+    auto actionName = externalName(actionDecl);
+    if (actionName != "NoAction") {
+        for (auto tcAction : tcPipeline->actionDefs) {
+            if (actionName == tcAction->actionName) {
+                tabledef->setDefaultMissAction(tcAction);
+                auto defaultActionProperty =
+                    t->properties->getProperty(IR::TableProperties::defaultActionPropertyName);
+                if (defaultActionProperty->isConstant) {
+                    tabledef->setDefaultMissConst(true);
+                }
+            }
+        }
+    }
+}
+
+void ConvertToBackendIR::updateDefaultHitAction(const IR::P4Table *t, IR::TCTable *tabledef) {
+    auto actionlist = t->getActionList();
+    if (actionlist != nullptr) {
+        unsigned int defaultHit = 0;
+        unsigned int defaultHitConst = 0;
+        cstring defaultActionName = nullptr;
+        for (auto action : actionlist->actionList) {
+            auto annoList = action->getAnnotations()->annotations;
+            bool isTableOnly = false;
+            bool isDefaultHit = false;
+            bool isDefaultHitConst = false;
+            for (auto anno : annoList) {
+                if (anno->name == "tableonly") {
```

Comment:
Simirarly here and below, can we use `ParseTCAnnotations::` ?

Reply:
updated the code